pkg/storage/stores/shipper/downloads/table_manager.go

```diff
@@ -0,0 +1,166 @@
+package downloads
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	pkg_util "github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const cacheCleanupInterval = 24 * time.Hour
+
+type Config struct {
+	CacheDir     string
+	SyncInterval time.Duration
+	CacheTTL     time.Duration
+}
+
+type TableManager struct {
+	cfg             Config
+	boltIndexClient BoltDBIndexClient
+	storageClient   StorageClient
+
+	tables    map[string]*Table
+	tablesMtx sync.RWMutex
+	metrics   *metrics
+
+	done chan struct{}
+	wg   sync.WaitGroup
+}
+
+func NewTableManager(cfg Config, boltIndexClient BoltDBIndexClient, storageClient StorageClient, registerer prometheus.Registerer) (*TableManager, error) {
+	return &TableManager{
+		cfg:             cfg,
+		boltIndexClient: boltIndexClient,
+		storageClient:   storageClient,
+		tables:          make(map[string]*Table),
+		metrics:         newMetrics(registerer),
+		done:            make(chan struct{}),
+	}, nil
+}
+
+func (tm *TableManager) loop() {
+	defer tm.wg.Done()
+
+	syncTicker := time.NewTicker(tm.cfg.SyncInterval)
+	defer syncTicker.Stop()
+
+	cacheCleanupTicker := time.NewTicker(cacheCleanupInterval)
+	defer cacheCleanupTicker.Stop()
+
+	for {
+		select {
+		case <-syncTicker.C:
+			err := tm.syncTables(context.Background())
+			if err != nil {
+				level.Error(pkg_util.Logger).Log("msg", "error syncing local boltdb files with storage", "err", err)
+			}
+		case <-cacheCleanupTicker.C:
+			err := tm.cleanupCache()
+			if err != nil {
+				level.Error(pkg_util.Logger).Log("msg", "error cleaning up expired tables", "err", err)
+			}
+		case <-tm.done:
+			return
+		}
+	}
+}
+
+func (tm *TableManager) Stop() {
+	close(tm.done)
+	tm.wg.Wait()
+
+	tm.tablesMtx.Lock()
+	defer tm.tablesMtx.Unlock()
+
+	for _, table := range tm.tables {
+		table.Close()
+	}
+}
+
+func (tm *TableManager) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback chunk_util.Callback) error {
+	return chunk_util.DoParallelQueries(ctx, tm.query, queries, callback)
+}
+
+func (tm *TableManager) query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	table := tm.getOrCreateTable(query.TableName)
+
+	// let us check if table is ready for use while also honoring the context timeout
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-table.IsReady():
+	}
+
+	if table.Err() != nil {
+		tm.tablesMtx.Lock()
+		defer tm.tablesMtx.Unlock()
+
+		delete(tm.tables, query.TableName)
```

Comment:
I think you should log table deletion like you're doing for addition. This would help debug if anything goes wrong.

Reply:
with the error that triggered the deletion may be ?

Reply:
Added the logs. Thanks!